File: Visual Studio LIghtswitch 2012/Chapter17/HelpDeskCS/HelpDeskCS/Client/UserCode/EditableIssuesGrid.cs
//You can use and redistribute the code from this book (and sample application) for non-commercial and 
//most commercial purposes as long as you acknowledge the source and authorship. 
//You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
//The acknowledgment should include author, title, publisher, and ISBN. 
//An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

using System;
using System.Linq;
using System.IO;
using System.IO.IsolatedStorage;
using System.Collections.Generic;
using Microsoft.LightSwitch;
using Microsoft.LightSwitch.Framework.Client;
using Microsoft.LightSwitch.Presentation;
using Microsoft.LightSwitch.Presentation.Extensions;
namespace LightSwitchApplication
{
    public partial class EditableIssuesGrid
    {

        //Listing 5-10. Validating Deletions on the Server 
        partial void EditableIssuesGrid_SaveError(Exception exception, ref bool handled)
        {
            ValidationException validationExc =
               (exception as ValidationException);
            if (validationExc != null)
            {

                var entities = from v in validationExc.ValidationResults
                               let e = (v.Target as IEntityObject)
                               where (e != null &&
                                  e.Details.EntityState == EntityState.Deleted)
                               select e;

                foreach (IEntityObject e in entities)
                {
                    e.Details.DiscardChanges();
                }
            }
        }
    }
}

File: Visual Studio LIghtswitch 2012/Chapter7/HelpDeskVB/HelpDeskVB/Client/UserCode/EditableIssuesGrid.vb
'You can use and redistribute the code from this book (and sample application) for non-commercial and 
'most commercial purposes as long as you acknowledge the source and authorship. 
'You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
'The acknowledgment should include author, title, publisher, and ISBN. 
'An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".


Namespace LightSwitchApplication

    Public Class EditableIssuesGrid

        ' Listing 5-10. Validating Deletions on the Server
        Private Sub EditableIssuesGrid_SaveError(
                                                exception As Exception, ByRef handled As Boolean)
            ' Un-delete deleted records that had server-side validation errors
            Dim validationExc = TryCast(exception, ValidationException)

            If validationExc IsNot Nothing Then
                Dim entities = From v In validationExc.ValidationResults
                   Let e = TryCast(v.Target, IEntityObject)
                     Where e IsNot Nothing AndAlso
                        e.Details.EntityState = EntityState.Deleted
                Select e
                For Each e In entities
                    e.Details.DiscardChanges()
                Next
            End If
        End Sub

    End Class

End Namespace
